### ISSUE TITLE
storage: experimental_relocate hangs

### ISSUE THREAD

**Issue by andreimatei** [cockroachdb/cockroach#31529]
On a 4-node cluster I've created a table with a few rows in it and split it into 4 ranges. Then I played with `split_at` and `experimental_relocate` and I got into the situation shown below (first query). Then, `experimental_relocate` reliably hangs:
```
root@:26257/t> SHOW EXPERIMENTAL_RANGES FROM TABLE orders;                                                                                                                             start_key | end_key | range_id | replicas | lease_holder
+-----------+---------+----------+----------+--------------+
  NULL      | /1      |     5613 | {1}      |            1
  /1        | /2      |     5615 | {2}      |            2
  /2        | /3      |     5616 | {1,2}    |            2
  /3        | NULL    |     5617 | {1,2,4}  |            2
(4 rows)

root@:26257/t> ALTER TABLE orders EXPERIMENTAL_RELOCATE                                                                                                                                SELECT ARRAY[1,2,3], 3;
```

**Comment by andreimatei** [comment]
Can repro with another table too. The 2nd `experimental_relocate` on the table seems to hang and not take effect.

**Comment by andreimatei** [comment]
correction, this is a 5 node cluster.
Repro below. The last relocate fails, and note that I'm asking it to put the lease on a node that doesn't currently have a replica.
```
root@:26257/t> create table orders3(id int primary key, customer_id int, amount int);
CREATE TABLE

Time: 49.991921ms

root@:26257/t> insert into orders3 values (1,1,1),(2,1,2),(3,2,1000),(4,2, 1500);
INSERT 4

Time: 34.267747ms

root@:26257/t> ALTER TABLE orders3 SPLIT AT SELECT i FROM generate_series(1, 9) as g(i);
      key      |    pretty
+--------------+---------------+
  \314\211\211 | /Table/68/1/1
  \314\211\212 | /Table/68/1/2
  \314\211\213 | /Table/68/1/3
  \314\211\214 | /Table/68/1/4
  \314\211\215 | /Table/68/1/5
  \314\211\216 | /Table/68/1/6
  \314\211\217 | /Table/68/1/7
  \314\211\220 | /Table/68/1/8
  \314\211\221 | /Table/68/1/9
(9 rows)

Time: 629.428034ms

root@:26257/t> show experimental_ranges from table orders3;
  start_key | end_key | range_id | replicas | lease_holder
+-----------+---------+----------+----------+--------------+
  NULL      | /1      |     5929 | {1,2,4}  |            2
  /1        | /2      |     5930 | {1,2,4}  |            2
  /2        | /3      |     5931 | {1,2,4}  |            2
  /3        | /4      |     5932 | {1,2,4}  |            2
  /4        | /5      |     5933 | {1,2,4}  |            2
  /5        | /6      |     5934 | {1,2,4}  |            2
  /6        | /7      |     5935 | {1,2,4}  |            2
  /7        | /8      |     5936 | {1,2,4}  |            2
  /8        | /9      |     5937 | {1,2,4}  |            2
  /9        | NULL    |     5938 | {1,2,4}  |            2
(10 rows)

Time: 16.46672ms

root@:26257/t> alter table orders3 experimental_relocate select array[4,2,1], 1;
      key      |    pretty
+--------------+---------------+
  \314\211\211 | /Table/68/1/1
(1 row)

Time: 16.275576ms

root@:26257/t> show experimental_ranges from table orders3;
  start_key | end_key | range_id | replicas | lease_holder
+-----------+---------+----------+----------+--------------+
  NULL      | /1      |     5929 | {1,2,4}  |            2
  /1        | /2      |     5930 | {1,2,4}  |            4
  /2        | /3      |     5931 | {1,2,4}  |            2
  /3        | /4      |     5932 | {1,2,4}  |            2
  /4        | /5      |     5933 | {1,2,4}  |            2
  /5        | /6      |     5934 | {1,2,4}  |            2
  /6        | /7      |     5935 | {1,2,4}  |            2
  /7        | /8      |     5936 | {1,2,4}  |            2
  /8        | /9      |     5937 | {1,2,4}  |            2
  /9        | NULL    |     5938 | {1,2,4}  |            2
(10 rows)

Time: 16.988292ms

root@:26257/t> alter table orders3 experimental_relocate select array[4,2,1], 1;
      key      |    pretty
+--------------+---------------+
  \314\211\211 | /Table/68/1/1
(1 row)

Time: 16.476864ms

root@:26257/t> alter table orders3 experimental_relocate select array[4,2,1], 2;
      key      |    pretty
+--------------+---------------+
  \314\211\212 | /Table/68/1/2
(1 row)

Time: 16.712583ms

root@:26257/t> show experimental_ranges from table orders3;
  start_key | end_key | range_id | replicas | lease_holder
+-----------+---------+----------+----------+--------------+
  NULL      | /1      |     5929 | {1,2,4}  |            2
  /1        | /2      |     5930 | {1,2,4}  |            4
  /2        | /3      |     5931 | {1,2,4}  |            4
  /3        | /4      |     5932 | {1,2,4}  |            2
  /4        | /5      |     5933 | {1,2,4}  |            2
  /5        | /6      |     5934 | {1,2,4}  |            2
  /6        | /7      |     5935 | {1,2,4}  |            2
  /7        | /8      |     5936 | {1,2,4}  |            2
  /8        | /9      |     5937 | {1,2,4}  |            2
  /9        | NULL    |     5938 | {1,2,4}  |            2
(10 rows)

Time: 14.407536ms

root@:26257/t>
root@:26257/t> alter table orders3 experimental_relocate select array[3,2,1], 2;
```

**Comment by a-robinson** [comment]
I've been throwing different variants of stores and primary keys at my `orders3` table for about 10 minutes now with no luck at getting anything to hang. I have encountered some poor behavior around what happens if you repeat the same store multiple times (sometimes it errors and sometimes it de-dupes them for you, and I've been able to get a couple "pq: change replicas of r%d failed: descriptor changed" errors (caused by #31287), but haven't managed to create any stalls.

If you can do it consistently, I'd love to sit next to you tomorrow and figure out what's going on. Or maybe it's a version difference? I'm on master (fd53070810620867d6ca45eb879b9031a36b838b).

**Comment by andreimatei** [comment]
Another day, more debugging; seems that the cluster in question (`andrei-tracing`) is pretty fucked. It would appear some ranges were badly merged and left in a state where the right hand side can't be GCed. It also appears at first glance that they were merged with ranges that were not adjacent (?!). In any case, @benesch is taking it over.
Unclear how the merge issues are causing the `experimental_relocate` command to hang, but we're in screwed up cluster territory.

**Comment by benesch** [comment]
Luckily they were actually adjacent:

```
 pre-merge r2328:/Table/57/1/461/10/2329/0-/Table/57/1/522/6/2815/PrefixEnd
 pre-merge r2326:/Table/57/1/522/6/2815/PrefixEnd-/Table/57/1/549/2/365
post-merge r2328:/Table/57/1/461/10/2329/0-/Table/57/1/549/2/365/0
```

**Comment by benesch** [comment]
The bug was in replica GC in the presence of merges. I pushed a patched build to the cluster and everything was resolved in less than a minute, including the `EXPERIMENTAL_RELOCATE` hang.

It's not surprising that the relocate got stuck, either; one of the replicas that got stuck spanned from something like [/Table/57/foo, /Max), so relocating range from orders3 (table ID 63) onto the store with the stuck replica was absolutely doomed.

Closing since I feel confident in this analysis. Follow #31409 for details, which describes the same issue.
